App deploy doc

ollieanwyll opened the pull request.

Commit:
Creation of app-deploy.md, Overview, Prereqs, Pushing to ECR

Commit:
First draft of deployment doc

Commit:
Minor formatting

alkar commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+Amazon's ECR, within ECS is where all of the application images used by the Cloud Platform are stored.
+
+### Creating or choosing a repository
```
Are the end-users supposed to have the necessary permissions to do this? If not, we could always manage this with terraform and PRs.

ollieanwyll replied:
One for @kalbir and @kerin I think.

alkar replied:
There's mention in [this doc](https://github.com/ministryofjustice/cloud-platform/blob/master/architecture-decision-record/007-Use-ECR-As-Container-Registry.md) on how user access can be implemented.

As far as I understand it's still WIP so I don't want to block this PR. I think it's fine to go as-is and we can revisit.

@kerin, thoughts?

kerin replied:
Good question - I think ultimately we want to allow users to interact with AWS directly in the console and on the command line, but that requires that we sort out AWS Federation and user credentials. Given that we're going to be handing S3, RDS etc. via Terraform it's probably simplest to do the same for ECR right now.

One thing to consider though - to stop this from getting blocked right now we could just create an IAM user with admin rights on ECR generally, and use those single set of creds across the board, and tackle per-image rights later - @kalbir @alkar ?

kalbir replied:
Any thoughts on how will we share that cred with the user who is trying to upload their image to ECR?

kerin replied:
We could just ping it around with gpg/Keybase/Signal or similar, but that might be pretty annoying. Maybe we could publish it in Confluence or a private repo, accepting that that is very much not the way to do it for real?

Alternative routes are

- obtain it from `terraform output` - but that requires that we tackle the AWS resources / terraform side of things
- have them use AWS Federation with github accounts - we're part of the way there, as we have the basic signin mechanism working, but we haven't yet addressed creating the IAM roles and policies users will assume

alkar replied:
Should we leave it as-is for now and wait for proper auth federation? It would mean that there's no way to actually follow the guide, however, if we expect auth to be sorted in the very near future I think it wouldn't be a real blocker.

Even if we temporarily took on the task of issuing a new pair of IAM creds for every potential user until auth is complete, would that be a lot of overhead? I'd like to avoid going the shared credentials route if possible.

kalbir replied:
Yep, let's go with the sharing a cred around right now. Next story off the backlog is for users to get access to AWS resources so we can make this bit a priority to work out. 

That means we will have to:

a) update this to say something along the lines of, "when you are uploading your image to ECR you'll need a credential from the cloud platform team, ask in the #cloud-platform-users channel in slack"
b) create a tightly scoped credential for this purposs -- allows a user to push to ECR and the resulting image is public so anyone can pull it. 
c) put the credential somewhere behind access control where we can share it

Agreed? @ollieanwyll are you happy to update the docs here with that?

kerin replied:
Having the Cloud Platform team create and issue creds for each user sounds good to me - we should probably do that via Terraform too so we keep track of what we've done in AWS.

tatyree replied:
On reflection, I'd *strongly* prefer we did this–start like we mean to go on...

ollieanwyll replied:
Amended with 83b8b27

alkar commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+On the ECR start page you will have the option create a new repository, or to search for an existing one.
+
+Create a new repository, and name it something relevant:
```
Should we instead/also provide instructions on how to do this on the CLI, like the steps below?

ollieanwyll replied:
Can do. I will have to learn how to do this myself first.

alkar commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+### Service
+
+Service files are used to specify port and protocol information for your application.
```
Should we also point out that Services are used to "bundle" a set of Pods created by the Deployment above?

alkar commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+### Ingress
+
+Ingress files are to use to specify host resolving information for your application.
```
I think this should read "[...] to define external access to the application". Technically host resolution would be DNS.

alkar commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+* **View pods** - `kubectl get pods`
+* **Check host** - `kubectl get ing`
+* **Delete application** - `kubectl delete deployment --all` + `kubectl delete service --all` + `kubectl delete ingress --all`
```
I think a cleaner/safer option is to just `kubectl delete -f deployment-files`.

Commit:
Minor wording changes based on Mitch feedback

kalbir commented:
PR for #2

tatyree commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+You will then be presented with a list of preconfigured terminal commands for you to run.
+
+Copy the first command. You will need to add `--profile yourAWSProfile` if the AWS account you're pushing to is not the default in your CLI:
+
```
You can also set the environment variable `AWS_ENVIRONMENT=<yourAWSProfile>`.

tatyree commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+Ensure the Docker image for your application has been built and is stored locally on your machine.
+
+Now we need to tag the image with the tag provided by ECR, so I can be pushed into the correct repository.
```
so **it** can be pushed...

tatyree commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+For the value of the `image:` key, you will see the Repository URI, which has been provided by ECR.
+
+If you are using the file as a template for your own application, replace the value of `image:` with the Repository URI of your application, found in ECR.
+
```
Is it worth putting a link here to the documentation for all of the options in a `deployment.yml` config?

kerin replied:
I'd include a link to the Kubernetes [deployment howto for developers](https://kubernetes.io/docs/tasks/run-application/run-stateless-application-deployment/).

ollieanwyll replied:
Amended with 83b8b27

tatyree commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,205 @@
+
+If you are using the file as a template for your own application, replace the `cp-demo-app-svc` tags with ones suited to your application.
+
+Also, ensure the `selector:` app tag is the same as specified in the `deployment.yml`.
```
Same comment as above, re. links to docs...

ollieanwyll replied:
Amended with 83b8b27

tatyree commented on cloud-platform/app-deploy.md:
````diff
@@ -0,0 +1,205 @@
+```
+
+If you are using the file as a template for your own application, replace the `cp-demo-app-ing` tag with one suited to your application.
+
````
ibid.

ollieanwyll replied:
Amended with 83b8b27

tatyree reviewed (changes requested):
The minor typo is the only thing I'll insist on being fixed here :)  Documentation links at your discretion.

Commit:
Amendments made based on PR feedback

ollieanwyll commented:
Feedback has been addressed.

kerin commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,214 @@
+
+Now ensure that the environment is empty by running each of the commands, `No resources found.` should be returned for all of them:
+
+* `kubectl get pods`
```
Do we need to mention specifying `--namespace` on all these `kubectl` commands? Or is `kubeconfig` and context setup with namespaces documented elsewhere?

kerin commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,214 @@
+
+With the environment empty, deploy the application by running the following command, that points to the `deployment-files` directory, in which the deployment files are stored:
+
+`kubectl create -f deployment-files`
```
It's probably better to refer to `apply` rather than `create` here, so we don't lead users into a `create` / `delete` cycle when making changes.

kerin commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,214 @@
+
+[https://github.com/ministryofjustice/cloud-platform-demo-app](https://github.com/ministryofjustice/cloud-platform-demo-app)
+
+### Deployment
```
I think it'd be useful to link to the Kubernetes docs for users when we mention k8s resources - in this case, [`Deployment`](https://kubernetes.io/docs/tasks/run-application/run-stateless-application-deployment/)

kerin commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,214 @@
+
+*Tip:* You can find more deployment config info [here.](https://kubernetes.io/docs/tasks/run-application/run-stateless-application-deployment/)
+
+### Service
```
https://kubernetes.io/docs/tasks/access-application-cluster/service-access-application-cluster/

kalbir replied:
Done in https://github.com/ministryofjustice/cloud-platform-user-docs/pull/18/commits/1dcb3083fcc0a032ad459ca642e94985b5304f2f

kerin commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,214 @@
+
+*Tip:* You can find more deployment config info [here.](https://kubernetes.io/docs/tasks/run-application/run-stateless-application-deployment/)
+
+### Ingress
```
https://kubernetes.io/docs/concepts/services-networking/ingress/

kalbir replied:
Done in https://github.com/ministryofjustice/cloud-platform-user-docs/pull/18/commits/1dcb3083fcc0a032ad459ca642e94985b5304f2f

kerin reviewed (dismissed):
Some general comments, and one bigger one - we don't really mention application environments or namespaces at all here - is that covered elsewhere?

razvan-moj-zz reviewed (approved):
LGTM

kalbir committed:
Make attaching to a namespace explicit

We assume you are deploying your application into the namespace that you
created as part of the [creating an environment]() guide but we did not
make the commands explicit enough.

Here we've decided to add `--namespace` as the approach as it makes it
explicit what you are doing and is a bit quicker for now than
configuring a context in your .kubeconfig. Happy for this to be changed
in the future.

kalbir committed:
Link to the kubernetes documentation

As suggested by @kerin and @taytree, we've linked to the kubernetes
documentation for the specific concepts we are talking about. In
particular this is ingress, service definitions and the general
deployment docs.

kalbir commented:
On @kerin's general comment about namespaces -- good point. We've made it explicit that you will have to reference the namespace that you created in the earlier documentation.

As per the commit, we've done this using `--namespace` for now but might be worth thinking about adding something that shows a user how to set it as a context (link to kube docs on that might be fine).

kerin commented on cloud-platform/app-deploy.md:
```diff
@@ -0,0 +1,212 @@
+* **View pods** - `kubectl get pods --namespace my-app-dev`
+* **Check host** - `kubectl get ing --namespace my-app-dev`
+* **Delete application** - `kubectl delete -f deployment-files --namespace my-app-dev`
+* **Shell into container** - `kubectl exec -it pod-name -- /bin/bash --namespace my-app-dev`
```
That'll error out I think - the command you're running in the container has to be at the end, so:

`kubectl exec -it --namespace my-app-dev pod-name -- /bin/bash`

kerin replied:
Also the long form of the `-it` flags might be a bit more informative?

`kubectl exec --stdin --tty --namespace my-app-dev pod-name -- /bin/bash`

kalbir replied:
Yes you are right 👍🏽

kalbir committed:
Fix exec command and make it more verbose

The fix and suggestion for making the command more verbose are both from
@Kerin. I think making commands like this more verbose is good in
general as we can't assume our users know the short versions.

Same for using `-n` vs `-namesapce`
